src/Interpreters/Cache/FileCache.cpp

```diff
@@ -681,23 +680,22 @@ bool FileCache::tryReserve(FileSegment & file_segment, size_t size)
 
     for (auto & [current_key, deletion_info] : to_delete)
     {
-        auto locked_key = deletion_info.getMetadata().tryLock();
+        auto locked_key = deletion_info.key_metadata->tryLock();
         if (!locked_key)
             continue; /// key could become invalid after we released the key lock above, just skip it.
 
-        for (auto it = deletion_info.begin(); it != deletion_info.end();)
+        for (const auto & candidate : deletion_info.candidates)
         {
-            chassert((*it)->releasable());
+            chassert(candidate->releasable());
 
-            auto segment = (*it)->file_segment;
+            auto segment = candidate->file_segment;
             locked_key->removeFileSegment(segment->offset(), segment->lock());
             segment->getQueueIterator()->remove(cache_lock);
 
             if (query_context)
                 query_context->remove(current_key, segment->offset(), cache_lock);
-
-            it = deletion_info.erase(it);
         }
+        deletion_info.candidates.clear();
```

Comment:
There was a removal of candidate after each iteration, because as file segment is successfully removed, then we do not need to iterate it in destructor of DeletionCandidates anymore. It was made for the case if there was an exception before or in the middle of iteration DeletionCandidates, so in such case we need to update state only for valid file segments, which have not yet been deleted. (Therefore for this scenario it was convenient to have inheritance from vector).

Reply:
Thanks for clarification, will change it